src/boot/efi/util.c

```diff
@@ -769,6 +770,114 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
         return EFI_SUCCESS;
 }
 
+#define QEMU_KERNEL_LOADER_FS_MEDIA_GUID \
+        { 0x1428f772, 0xb64a, 0x441e, {0xb8, 0xc3, 0x9e, 0xbd, 0xd7, 0xf8, 0x93, 0xc7 }}
+
+#define QEMU_BOOT_ORDER_GUID \
+        { 0x668f4529, 0x63d0, 0x4bb5, {0xb6, 0x5d, 0x6f, 0xbb, 0x9d, 0x36, 0xa4, 0x4a }}
+
+/* detect direct kernel boot, i.e. 'qemu -kernel systemd-bootx64.efi' */
+bool is_qemu_direct_kernel(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return false;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH ||
+            dp->Header.SubType != MEDIA_VENDOR_DP)
+                return false;
+
+        if (memcmp(&dp->Guid, &(EFI_GUID)QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) != 0)
+                return false;
+
+        return true;
+}
+
+/* detect loading from firmware volume (sd-boot added to ovmf) */
+bool is_firmware_volume(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return false;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH ||
+            dp->Header.SubType != MEDIA_PIWG_FW_VOL_DP)
+                return false;
+
+        return true;
+}
+
+/* try find ESP when not loaded from ESP */
+EFI_STATUS qemu_open(EFI_HANDLE *qemu_dev, EFI_FILE **qemu_dir) {
```

Comment:
We usually prefix `ret_` for arguments that store function results.
Also, please add assertions for the arguments.